include/graaflib/algorithm/coloring/welsh_powell.tpp

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <graaflib/algorithm/coloring/welsh_powell.h>
+
+#include <unordered_map>
+#include <vector>
+#include <algorithm>
+#include <iostream>
+
+namespace graaf::algorithm {
+
+template <typename GRAPH>
+std::unordered_map<vertex_id_t, int> welsh_powell_coloring(const GRAPH& graph) {
+    
+    using DegreeVertexPair = std::pair<int, vertex_id_t>;
+
+    // Step 1: Sort vertices by degree in descending order
+    std::vector<DegreeVertexPair> degree_vertex_pairs;
+    for (const auto& vertex : graph.get_vertices()) {
+        int degree = graph.get_degree(vertex);
+        degree_vertex_pairs.emplace_back(degree, vertex);
+    }
+
+    std::sort(degree_vertex_pairs.rbegin(), degree_vertex_pairs.rend());
+
+    // Step 2: Assign colors to vertices
+    std::unordered_map<vertex_id_t, int> color_map;
```

Comment:
I see there is currently an issue in the other coloring algorithm (`greedy_graph_coloring`) where the algorithm does not assign a "minimal" coloring. If it did, we could extract the logic for assigning the colors to a common function. Maybe something with a signature such as:

```c++
template <typename GRAPH>
[[nodiscard]] std::unordered_map<vertex_id_t, int> assign_colors(const std::vector<vertex_id_t>& vertices, const GRAPH& graph) {...}
```

Then we could pass in the (non-sorted) vertex IDs for the greedy algorithm and the sorted ones for Welsh-Powell.

What do you think? Might be a nice follow-up

Reply:
This would indeed be a great follow up but can we take it up in a different PR?

Reply:
Indeed, I would definitely do this is a separate PR 👍🏻